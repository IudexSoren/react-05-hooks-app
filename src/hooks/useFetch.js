import { useEffect, useRef, useState } from 'react';

export const useFetch = (url) => {

  const isMounted = useRef(true);
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null
  });

  useEffect(() => {
    return () => {
      isMounted.current = false;
    }
  }, [])

  useEffect(() => {
    setState({data: null, loading:true, error: null});
    fetch(url)
    .then((resp) => resp.json())
    .then( async(data) => {
      // isMounted evita error por componente desmontado
      if (isMounted.current) {
        setState({
          data: data,
          loading: false,
          error: null
        });
      }
    })
    .catch(() => {
      setState({
        data: null,
        loading: false,
        error: 'La informaci√≥n no pudo ser cargada'
      });
    });
  }, [url]);

  return state;
}