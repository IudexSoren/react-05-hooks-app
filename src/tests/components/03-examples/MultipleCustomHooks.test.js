import { shallow } from 'enzyme';
import MultipleCustomHook from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas de MultipleCustomHook', () => {

  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 0,
      increment: () => {}
    });
  });

  test('Debe renderizar el componente correctamente', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    });
    const wrapper = shallow(<MultipleCustomHook />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Debe mostrar la informaciÃ³n', () => {
    useFetch.mockReturnValue({
      data: [{
        author: 'Putando',
        quote: 'Que buenas tetas'
      }],
      loading: false,
      error: null
    });
    const wrapper = shallow(<MultipleCustomHook />);

    expect(wrapper.find('.alert').exists()).toBeFalsy();
    expect(wrapper.find('.mb-3').text().trim()).toBe('Que buenas tetas');
    expect(wrapper.find('.blockquote-footer').text().trim()).toBe('Putando');
  });


});