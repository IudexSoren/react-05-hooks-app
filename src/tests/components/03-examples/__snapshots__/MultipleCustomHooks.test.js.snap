// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de MultipleCustomHook Debe renderizar el componente correctamente 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-outline-dark"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
